v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java

```diff
@@ -0,0 +1,451 @@
+   * pipeline is finished before returning. Invoke {@code result.waitUntilFinish()} on the result
+   * object to block until the pipeline is finished running if blocking programmatic execution is
+   * required.
+   *
+   * @param options The execution options.
+   * @return The pipeline result.
+   */
+  public static PipelineResult run(KafkaToBQFlexOptions options)
+      throws IOException, RestClientException {
+
+    // Validate BQ STORAGE_WRITE_API options
+    BigQueryIOUtils.validateBQStorageApiOptionsStreaming(options);
+    MetadataValidator.validate(options);
+
+    // Create the pipeline
+    Pipeline pipeline = Pipeline.create(options);
+
+    List<String> topicsList;
+    if (options.getKafkaReadTopics() != null) {
+      topicsList = new ArrayList<>(Arrays.asList(options.getKafkaReadTopics().split(",")));
+    } else {
+      throw new IllegalArgumentException("Please Provide --kafkaReadTopic");
+    }
+    String bootstrapServers;
+    if (options.getReadBootstrapServers() != null) {
+      bootstrapServers = options.getReadBootstrapServers();
+    } else {
+      throw new IllegalArgumentException("Please Provide --bootstrapServers");
+    }
+    /*
+     * Steps:
+     *  1) Read messages in from Kafka
+     *  2) Transform the messages into TableRows
+     *     - Transform message payload via UDF
+     *     - Convert UDF result to TableRow objects
+     *  3) Write successful records out to BigQuery
+     *  4) Write failed records out to BigQuery
+     */
+
+    Map<String, Object> kafkaConfig =
+        ImmutableMap.of(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, options.getKafkaReadOffset());
+
+    String username = SecretManagerUtils.getSecret(options.getKafkaReadUsernameSecretId());
+    String password = SecretManagerUtils.getSecret(options.getKafkaReadPasswordSecretId());
+
+    ImmutableMap<String, Object> finalKafkaConfig =
+        ImmutableMap.<String, Object>builder()
+            .putAll(kafkaConfig)
+            .putAll(setClientAuthConfig(username, password))
+            .build();
+
+    if (options.getMessageFormat() == null || options.getMessageFormat().equals("JSON")) {
+
+      return runJsonPipeline(pipeline, options, topicsList, bootstrapServers, finalKafkaConfig);
+
+    } else if (options.getMessageFormat() == null || options.getMessageFormat().equals("AVRO")) {
+
+      return runAvroPipeline(pipeline, options, topicsList, bootstrapServers, finalKafkaConfig);
+
+    } else {
+      throw new IllegalArgumentException("Invalid format specified: " + options.getMessageFormat());
+    }
+  }
+
+  public static PipelineResult runAvroPipeline(
+      Pipeline pipeline,
+      KafkaToBQFlexOptions options,
+      List<String> topicsList,
+      String bootstrapServers,
+      Map<String, Object> kafkaConfig)
+      throws IOException, RestClientException {
+
+    if (options.getAvroFormat().equals("NON_WIRE_FORMAT") && options.getAvroSchemaPath() == null) {
```

Comment:
Can be fixed later during refactoring: this source format should share the same enum with KafkaToGcs template.